cohorts/load.py

```diff
@@ -1020,6 +1020,93 @@ def plot_joint(self, on, on_two=None, **kwargs):
         p = sb.jointplot(data=df, x=plot_cols[0], y=plot_cols[1])
         return p
 
+    def summarize_provenance(self):
+        """ Utility function to summarize provenance of Cohort data for each existing cache_dir
+            
+            == Returns == 
+
+            Dict containing unique provenance among all patients in each existing cache dir, or 
+            None if provenance is not unique among all patients.
+
+            Dict also contains an item 'dfhash' representing unique hash of default dataframe for cohort, given options.
+            
+            == Behavior == 
+
+            Checks for non-equivalent provenance among patients in each cache dir that exists. Cache dirs that
+            do not exist are not populated in the dictionary.
+
+            If non-equivalence is found:
+               - prints a warning
+               - returns None for that cache dir in the resulting dict
+
+            Otherwise:
+               - returns unique provenance for that cache dir, equivalent among all patients in data frame
+            
+            == See also ==
+            
+            `?cohorts.summarize_unique_provenance` which optionally prints a single provenance dict if provenance is identical across all
+            cache_dirs.
+
+        """
+        provenance_summary = {}
+        df = self.as_dataframe()
+        df_hash = hash(str(df.sort_values('patient_id')))
+        for cache in self.cache_names:
+            cache_name = self.cache_names[cache]
+            cache_provenance = None
+            cache_warnings = ""
+            this_cache_dir = path.join(self.cache_dir, cache_name)
+            if path.exists(this_cache_dir):
+                for i, row in df.iterrows():
+                    patient_id = row["patient_id"]
+                    patient_cache_dir = path.join(this_cache_dir, patient_id)
+                    this_provenance = self.load_provenance(patient_cache_dir = patient_cache_dir)
+                    if not(cache_provenance):
+                        cache_provenance = this_provenance
+                    else:
+                        cache_warnings += _compare_provenance(this_provenance, cache_provenance)
+                if len(cache_warnings) == 0:
+                    provenance_summary[cache_name] = cache_provenance
+                else:
+                    provenance_summary[cache_name] = None
+        provenance_summary[u'dfhash'] = df_hash
```

Comment:
I'm not too familiar with when/when not to use the term `provenance`; I imagine that it is appropriate to describe the hash. My concern is more around not treating the hash like a cache or a package. So maybe:

````
Provenance:

DataFrame hash: ___
Packages: 
{cohorts: ...}
```

I guess you could create a `Provenance` object that contains a package `dict` and a `df_hash`, a nd the `__str__` is the formatted string above?